samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -242,74 +150,105 @@ public boolean waitForFinish(Duration timeout) {
     return finished;
   }
 
+  @VisibleForTesting
+  protected Set<StreamProcessor> getProcessors() {
+    return Collections.unmodifiableSet(processors);
+  }
+
+  @VisibleForTesting
+  CountDownLatch getShutdownLatch() {
+    return shutdownLatch;
+  }
+
+  /* package private */
+  StreamProcessor createStreamProcessor(Config config, ApplicationDescriptorImpl<? extends ApplicationDescriptor> appDesc,
+      StreamProcessor.StreamProcessorListenerFactory listenerFactory) {
+    TaskFactory taskFactory = TaskFactoryUtil.getTaskFactory(appDesc);
+    Map<String, MetricsReporter> reporters = new HashMap<>();
+    // TODO: the null processorId has to be fixed after SAMZA-1835
+    appDesc.getMetricsReporterFactories().forEach((name, factory) ->
+        reporters.put(name, factory.getMetricsReporter(name, null, config)));
+    return new StreamProcessor(config, reporters, taskFactory, listenerFactory, null);
+  }
+
   /**
-   * Create intermediate streams using {@link org.apache.samza.execution.StreamManager}.
-   * If {@link CoordinationUtils} is provided, this function will first invoke leader election, and the leader
-   * will create the streams. All the runner processes will wait on the latch that is released after the leader finishes
-   * stream creation.
-   * @param planId a unique identifier representing the plan used for coordination purpose
-   * @param intStreams list of intermediate {@link StreamSpec}s
+   * Defines a specific implementation of {@link ProcessorLifecycleListener} for local {@link StreamProcessor}s.
    */
-  private void createStreams(String planId,
-      List<StreamSpec> intStreams,
-      StreamManager streamManager) {
-    if (intStreams.isEmpty()) {
-      LOG.info("Set of intermediate streams is empty. Nothing to create.");
-      return;
+  final class LocalStreamProcessorLifecycleListener implements ProcessorLifecycleListener {
+    private final StreamProcessor processor;
+    private final ProcessorLifecycleListener processorLifecycleListener;
```

Comment:
Could you please name this something like "innerProcessorLifecycleListener" or "delegateProcessorLifecycleListener"? I think it might help readability since there is this top-level LocalStreamProcessorLifecycleListener and then also the one that the user provides.

Reply:
Good point. I think that userDefinedListener is more straightforward. Let me know if you have a better name for it.